# pythonにおける変数 ～ミュータブルとイミュータブル～

## 1 ミュータブルとイミュータブルとは
- **ミュータブル** ... 変更可能な変数の型
- **イミュータブル** ... 変更不可能な変数の型（static変数)

### 1.1 pythonにおけるミュータブルとイミュータブルの例
- ミュータブルである変数型の例 ... 数値型(int, float)、文字列型、タプル型など
- イミュータブルである変数型の例 ... ディクショナリ型、リスト型

## 2 イミュータブルだけど見かけはミュータブル
pythonで、数値や文字列を扱うときはイミュータブルであることを意識しなくとも簡単に利用することができる。    
例えば、以下の処理には何の不思議もない。

In [3]:
x = 100
x = 100*3
x

300

しかし、xという変数はイミュータブルのに、なぜこのような処理ができるのか。    
それはアドレスを確認することで、わかる。

In [7]:
x = 100
print(id(x))
x = x*2
print(id(x))
x

1980870776272
1980870779536


200

xが保持しているアドレス(ID)が1980870776272から1980870779536に変わっている。   
これを、スタック領域とヒープ領域で考えると以下のように図示できる。
```
スタック：x => 1980870776272
ヒープ　：1980870776272 => 100
x = x*2
スタック：旧x => 1980870776272
　　　　　新x => 1980870779536
ヒープ　：1980870776272 => 100
　　　　　1980870779536 => 200
```

C言語の処理と対比してみる。
```
int a = 1;
a = a*100;
```
C言語において、値の更新の前後で変数aのスタック領域において格納している

In [9]:
x = 123
print(id(x))
x = 'abc'
print(id(x))

1980870777008
1980876006192
